### PR TITLE
Implement Einsum reference in nGraph interpreter

### PR DESCRIPTION
**Pull request by rkazants**
**Description:** 
1. It implements *Einsum-7* reference. The implementation is added into nGraph interpreter due to huge size of the implementation. The interpreter binary size (ngraph_reference lib) increases from 2414KB to 3292KB by **878KB** (or **36%**). Since plugins do not support *Einsum-7* operation directly, there exists *EinsumDecomposition* transformation that helps to complete all model requests.
2. This *Einsum* reference does not have limitations and supports all equations including repeated labels in input subscribt (aka diagonal extraction), ellipsis, and others.
3. The reference was tested using 11 unit tests with simple input data and ones generated with numpy

**Ticket:** 49327

**Details:** The main idea was inherited from *EinsumDecomposition* transformation implementation in [PR5529](https://github.com/openvinotoolkit/openvino/pull/5529) and extended to support diagonal extraction and ellipsis in an equation. To get the idea about support of equations without diagonal extraction and without ellipsis, please refer to [PR5529](https://github.com/openvinotoolkit/openvino/pull/5529) details where it describes the procedure to split labels into three groups (common, separated, reduced dimensions), permute them for grouping, apply *MatMul* operation, and reshape back the resulted tensor with transpose to get layout adjusted with output subscript.
Here it describes two cases: diagonal extraction and ellipsis.
**EinsumDiagonal**. For example, *Einsum* with `equation="abb->a"` will extract diagonals per batch (i.e. zero dimension):
image
It makes sense to note that diagonal extraction must be performed in the first step for each operand in case repeated labels in the corresponding input subscripts. For example, diagonal must be extracted for the first operand of *Einsum* with `equation="abbac,bd->ad"` in the first step. Only after the first step you can proceed with handling ellipsis and details from [PR5529](https://github.com/openvinotoolkit/openvino/pull/5529). 
Let us perform diagonal extraction for the first operand with input subscript `abbac` and shape equal to `[2, 3, 3, 2, 4]`.
1. For this, determine repeated labels and corresponding dimension indices for them. In this case, `a` and `b` are repeated labels and `[0, 3]` and `[1, 2]` are the corresponding dimension indices.
2. Build identity tensors for each repeated label. For `a` label, it will be `I1` of shape `[2,1,1,2,1]` where `I1[k,*,*,k,*]=1` for `k=0,1` and 0 - otherwise. For `b` label, it will be `I2` of shape `[1,3,3,1,1]` where `I2[k,*,*,k,*]=1` for `k=0,1,2` and 0 - otherwise. Note that identity rank matches to the input rank and its dimension sizes equal dimension sizes corresponding to given label from input shape and equal to 1 for all other labels.
3. Multiples all identities by element-wise operation `I=I1*I2*...*In` where `n` is a number of repeated labels.
4. Multiples the first operand by `I` and perform reduce-sum along dimensions from the corresponding dimension indices for repeated labels excluding the first dimension, i.e. `[3]` for `a` and `[2]` for `b`. Hence, the extracted diagonal will be of shape `[2,3,4]` with subscript `abc`.

**Ellipsis**. In case of ellipsis, the procedure described in [PR5529](https://github.com/openvinotoolkit/openvino/pull/5529) is extended with an ellipsis case. All steps from [PR5529](https://github.com/openvinotoolkit/openvino/pull/5529) are modified so that it takes into account that one ellipsis label can cover multiple dimension. Also, before *MatMul* operation you need to broadcast operands if they both have dimensions covered by ellipsis label.

Signed-off-by: Roman Kazantsev <roman.kazantsev@intel.com>

### CODE REVIEW

**Comment by lazarevevgeny** [comment]
Please, add information about the binary size increase of the ngraph library because of this operation.

**Comment by ilyachur** [comment]
Could you set the milestone for this activity?
Does this PR fix some ticket?

**Comment by rkazants** [comment]
> Could you set the milestone for this activity?
> Does this PR fix some ticket?

Done.

**Comment by rkazants** [comment]
> Please, add information about the binary size increase of the ngraph library because of this operation.

Done

File: ngraph/core/reference/src/runtime/reference/einsum.cpp [hunk 1]
```diff
@@ -0,0 +1,1134 @@
...
+            new HostTensor(input_ptr->get_element_type(), output_shape));
+
+        ngraph::runtime::reference::sum<T>(input_ptr->get_data_ptr<T>(),
+                                           output_ptr->get_data_ptr<T>(),
+                                           input_shape,
+                                           reduced_axes,
+                                           false);
+
+        // update a vector of inputs and input subscripts
+        inputs[input_ind] = output_ptr;
```
Comment (rkazants) [inline ngraph/core/reference/src/runtime/reference/einsum.cpp:411]: @ilyachur, please pay attention to memory management in this implementation. Now I create separate `HostTensor` for each intermediate result. `HostTensor` provides convenient way to incapsulate memory pointer and shape that I need for this implementation, however, it does not provide a convenient way to reuse it for another tensor with a new shape.

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 2]
```diff
@@ -0,0 +1,1210 @@
...
+                    transpose_input<T>(int_inputs, input_subscripts, output_subscript, 0);
+
+                    // write the result into output host tensor
+                    auto output_shape = int_inputs[0]->get_shape();
+                    const auto& element_type = int_inputs[0]->get_element_type();
+                    const auto& buf_size = shape_size(output_shape) * element_type.size();
+                    outputs[0]->write(int_inputs[0]->get_data_ptr(), buf_size);
+                }
+
+            } // namespace einsum_details
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:1185]: I propose to move all details to `cpp` file.

Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:1185]: I have already done so but I encountered build issue about multiple defined `runtime::transpose` function in unity configuration. This approach with hpp seems working

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 3]
```diff
@@ -0,0 +1,1210 @@
...
+            namespace einsum_details
+            {
+                std::vector<std::pair<size_t, size_t>> compute_einsum_path(size_t num_inputs)
+                {
+                    // TODO: implement algorithm for finding (pseudo-)optimal einsum_path
+                    std::vector<std::pair<size_t, size_t>> einsum_path;
+                    NGRAPH_CHECK(num_inputs > 0);
+                    for (size_t input_ind = num_inputs - 1; input_ind > 0; --input_ind)
+                    {
+                        einsum_path.push_back(std::make_pair(0, input_ind));
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:33]: is it correct that `einsum_path.size()` will be equal `num_inputs - 1`?

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 4]
```diff
@@ -0,0 +1,1210 @@
...
+                            return false;
+                        }
+                        ++check_index;
+                    }
+                    return true;
+                }
+
+                bool is_dimension_reduced(const std::vector<std::string>& input_subscripts,
+                                          const std::string& output_subscript,
+                                          const std::string label_to_check,
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:54]:
```suggestion
                                          const std::string& label_to_check,
```

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 5]
```diff
@@ -0,0 +1,1210 @@
...
+                        std::find(labels.begin(), labels.end(), ellipsis) != labels.end())
+                    {
+                        NGRAPH_CHECK(
+                            false,
+                            "Input rank cannot be dynamic in case of ellipsis in input subscript");
+                    }
+                    else if (input_rank.is_static())
+                    {
+                        input_rank_length = input_rank.get_length();
+                    }
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:136]:
```suggestion
                    NGRAPH_CHECK(
                            input_rank.is_static() || std::find(labels.begin(), labels.end(), ellipsis) == labels.end(),
                            "Input rank cannot be dynamic in case of ellipsis in input subscript");
                   if (input_rank.is_static())
                    {
                        input_rank_length = input_rank.get_length();
                    }
```

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 6]
```diff
@@ -0,0 +1,1210 @@
...
+                                    size_t& common_begin,
+                                    size_t& common_end,
+                                    size_t& sep_begin,
+                                    size_t& sep_end,
+                                    size_t& reduced_begin,
+                                    size_t& reduced_end,
+                                    bool is_separated_first)
+                {
+                    auto label_to_dim_map = compute_label_dim_map(input_rank, input_subscript);
+                    const std::string ellipsis = "...";
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:211]: it can be constexpr

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 7]
```diff
@@ -0,0 +1,1210 @@
...
+                                    contract_output,
+                                    resultant_subscript);
+                }
+
+                template <typename T>
+                void einsum_impl(const HostTensorVector& inputs,
+                                 const HostTensorVector& outputs,
+                                 const std::string& equation)
+                {
+                    // parse equation to extract input subscripts and output subscript
```
Comment (pavel-esir) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:1144]: is this comment really needed? I propose to remove this line. Code a line above in my opinion is quite self-explanatory

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 8]
```diff
@@ -0,0 +1,1210 @@
...
+                    // extract diagonal for the single operand
+                    extract_diagonal<T>(int_inputs, input_subscripts, 0);
+
+                    // reduce dimensions for the remained input node
+                    reduce_input<T>(int_inputs, input_subscripts, output_subscript, 0);
+
+                    // transpose dimensions to layout required by the output subscript
+                    transpose_input<T>(int_inputs, input_subscripts, output_subscript, 0);
+
+                    // write the result into output host tensor
```
Comment (pavel-esir) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:1178]: nit: i think this comment is redundant, it's pretty clear from code below what it does

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 9]
```diff
@@ -0,0 +1,1210 @@
...
+                            label_dims.push_back(current_dim);
+                            resulted_map[label] = label_dims;
+                            ++current_dim;
+                        }
+                    }
+
+                    return resulted_map;
+                }
+
+                Shape compute_sub_shape(const Shape& input_shape,
```
Comment (pavel-esir) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:172]: Is it possible to reuse `ngraph::shape` routine here?
https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/ngraph/shape.hpp#L23

Example of using is here https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/gather.hpp#L32-L33

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 10]
```diff
@@ -0,0 +1,1210 @@
...
+                        return Shape();
+                    }
+                    Shape sub_shape(input_shape.begin() + begin, input_shape.begin() + end);
+
+                    if (is_product)
+                    {
+                        size_t prod = 1;
+                        for (auto dim : sub_shape)
+                        {
+                            prod *= dim;
```
Comment (pavel-esir) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:189]: looks like in lines 184-190 can be replaced by `reference::span` and `ngraph::shape` routines https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/utils/span.hpp#L67 
 
https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/ngraph/shape.hpp#L23

example of using them here
https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/gather.hpp#L32-L33

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 11]
```diff
@@ -0,0 +1,1210 @@
...
+                    reduce_input<T>(inputs, input_subscripts, output_subscript, input_ind1);
+                    reduce_input<T>(inputs, input_subscripts, output_subscript, input_ind2);
+
+                    // step 0. split dimensions of both operands into three groups:
+                    // 1. dimension indices with the same labels (in both subscripts) that are NOT
+                    // reduced - common labels (dimensions)
+                    // 2. dimension indices with labels that are met only in one of two subscripts -
+                    // separate labels (dimensions)
+                    // 3. dimension indices with the same labels (in both subscripts) that are
+                    // reduced - reduced labels (dimensions) NOTE: dimension is reduced iff. the
```
Comment (pavel-esir) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:844]: `dimension is reduced iff` a typo?

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 12]
```diff
@@ -0,0 +1,1210 @@
...
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            namespace einsum_details
+            {
+                std::vector<std::pair<size_t, size_t>> compute_einsum_path(size_t num_inputs)
```
Comment (popovaan) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:26]: This function looks similar to compute_einsum_path() from einsum_decomposition.cpp. Can a single function be used?

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp [hunk 13]
```diff
@@ -0,0 +1,1210 @@
...
+                        if (check_index != index)
+                        {
+                            return false;
+                        }
+                        ++check_index;
+                    }
+                    return true;
+                }
+
+                bool is_dimension_reduced(const std::vector<std::string>& input_subscripts,
```
Comment (popovaan) [inline ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:52]: This function is similar to is_dimension_reduced() from einsum_decomposition.cpp. Can it be reused here?

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 3 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:33]: Exactly correct because the number of operand pairs `num_inputs - 1`

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 11 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:844]: iff = if and only if:). But anyway, I will re-write it to `if`

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 12 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:26]: Absolutely, agree with you. The situation is not trivial here because we have no common utils routine for both transformations and nGraph core. I have already tried this way but failed. The duplication including three function (`compute_einsum_path`, `is_dimension_reduced`, etc) is about 100 lines not too much. So let us keep these duplications.
Regarding the way to include them into nGraph operation API I consider that `compute_einsum_path` is not finalized (since I mark with TODO) in terms of its API and implementation so we are not ready to include this into nGraph op API.

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 13 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp:52]: Absolutely, agree with you. The situation is not trivial here because we have no common utils routine for both transformations and nGraph core. I have already tried this way but failed. The duplication including three function (compute_einsum_path, is_dimension_reduced, etc) is about 100 lines not too much. So let us keep these duplications.
Regarding the way to include `is_dimension_reduced` into nGraph op API I consider that this function is very specific to implementation and it is not likely that plugin will use this way.

**Review by popovaan — APPROVED** [review]

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 4 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp]: Fixed.

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 5 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp]: Done.

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 8 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp]: Removed.

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 9 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp]: I did changes inside this function. I only managed to use `shape_size` function. I can't assign `Shape` object with `Span` value.

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 10 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp]: I did changes inside this function. I only managed to use shape_size function. I can't assign Shape object with Span value. All lines with `prod` computation are now changed to a way with `shape_size`

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 11 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp]: Fixed.

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 6 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp]: Fixed.

File: ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp (hunk 7 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp]: Fixed.

**Review by pavel-esir — APPROVED** [review]

File: ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp [hunk 14]
```diff
@@ -56,7 +56,7 @@ namespace ngraph
                     }
                 }
 
-                std::vector<size_t> get_transpose_order(const Shape& input_shape)
+                static std::vector<size_t> get_transpose_order(const Shape& input_shape)
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp:59]: why did you add static to this function? I propose to revert this change

File: ngraph/test/op_eval/einsum.cpp [hunk 15]
```diff
@@ -0,0 +1,237 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <string>
```
Comment (ilyachur) [inline ngraph/test/op_eval/einsum.cpp:5]: Please add new tests for Visitor API.
https://github.com/openvinotoolkit/openvino/tree/master/ngraph/test/visitors/op

File: ngraph/core/reference/src/runtime/reference/einsum.cpp [hunk 16]
```diff
@@ -0,0 +1,1239 @@
...
+                }
+
+            } // namespace einsum_details
+
+            void einsum(const HostTensorVector& outputs,
+                        const HostTensorVector& inputs,
+                        const std::string& equation)
+            {
+                NGRAPH_CHECK(inputs.size() > 0, "Einsum must accept at least one input.");
+                auto input_type = inputs[0]->get_element_type();
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/einsum.cpp:1220]: Do we sure that all inputs have the same element type?

File: ngraph/core/reference/src/runtime/reference/einsum.cpp [hunk 17]
```diff
@@ -0,0 +1,1239 @@
...
+
+                    HostTensorVector int_inputs = inputs;
+
+                    // contract inputs by Einsum until just one is remained
+                    for (auto const& inds_pair : einsum_path)
+                    {
+                        contract_two_inputs<T>(int_inputs,
+                                               input_subscripts,
+                                               output_subscript,
+                                               inds_pair.first,
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/einsum.cpp:1192]: As I understand `inds_pair.first` is always equal to `0`, do we need to have this pair?

File: ngraph/core/reference/src/runtime/reference/einsum.cpp [hunk 18]
```diff
@@ -0,0 +1,1239 @@
...
+                template <typename T>
+                void contract_two_inputs(HostTensorVector& inputs,
+                                         std::vector<std::string>& input_subscripts,
+                                         const std::string& output_subscript,
+                                         size_t input_ind1,
+                                         size_t input_ind2)
+                {
+                    // assume that input_ind1 < input_ind2 without loss of generality, otherwise,
+                    // just swap them
+                    if (input_ind2 < input_ind1)
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/einsum.cpp:844]: Looks like `input_ind1` is always `0`

File: ngraph/core/reference/src/runtime/reference/einsum.cpp [hunk 19]
```diff
@@ -0,0 +1,1239 @@
...
+                    auto repeated_label_dims = label_dim_map[first_repeated_label];
+                    HostTensorPtr multi_identity =
+                        build_identity<T>(input_ptr, repeated_label_dims);
+
+                    for (size_t label_ind = 1; label_ind < repeated_labels.size(); ++label_ind)
+                    {
+                        NGRAPH_CHECK(label_dim_map.find(repeated_labels[label_ind]) !=
+                                     label_dim_map.end());
+                        repeated_label_dims = label_dim_map[repeated_labels[label_ind]];
+                        HostTensorPtr identity = build_identity<T>(input_ptr, repeated_label_dims);
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/einsum.cpp:653]: Looks like this `HostTensor` should be equal to `multi_identity` from 645 line. Can we reuse tensors which contain same data?

File: ngraph/core/reference/src/runtime/reference/einsum.cpp [hunk 20]
```diff
@@ -0,0 +1,1239 @@
...
+
+                        PartialShape output_shape = multi_identity->get_partial_shape();
+                        PartialShape::broadcast_merge_into(output_shape,
+                                                           identity->get_partial_shape(),
+                                                           ngraph::op::AutoBroadcastSpec::NUMPY);
+                        HostTensorPtr mul_output = std::shared_ptr<HostTensor>(
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/einsum.cpp:659]: In any case here you create one more tensor for output data.

File: ngraph/core/reference/src/runtime/reference/einsum.cpp [hunk 21]
```diff
@@ -0,0 +1,1239 @@
...
+                        ngraph::runtime::reference::multiply<T>(
+                            unsqueeze_output1->get_data_ptr<T>(),
+                            unsqueeze_output2->get_data_ptr<T>(),
+                            mul_output->get_data_ptr<T>(),
+                            unsqueeze_output1->get_shape(),
+                            unsqueeze_output2->get_shape(),
+                            ngraph::op::AutoBroadcastSpec::NUMPY);
+
+                        // update input operand and input subscript for Einsum operation
+                        update_operands(inputs,
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/einsum.cpp:1005]: If you update inputs, looks like in some method which were called before, we can rewrite data in input tensors instead of new tensor creation.

File: ngraph/core/reference/src/runtime/reference/einsum.cpp [hunk 22]
```diff
@@ -0,0 +1,1239 @@
...
+                    }
+                    if (input_shape == result_shape)
+                    {
+                        return;
+                    }
+
+                    HostTensorPtr multi_identity =
+                        build_multi_identity<T>(input_ptr, repeated_labels, label_dim_map);
+
+                    HostTensorPtr mul_output = std::shared_ptr<HostTensor>(
```
Comment (ilyachur) [inline ngraph/core/reference/src/runtime/reference/einsum.cpp:733]: Looks like here we can use `input_ptr` as destination.

File: ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp (hunk 14 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp:59]: Otherwise, I received names conflict during linking. The same name is declared for transformations. This function must be internal

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 17 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp:1192]: Yes, we do because `compute_einsum_path` is now undone. And it will be complete once we receive model request with several operand Einsum.

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 18 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp:844]: Now yes, but someday it can be changed once `compute_einsum_path` is complete.

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 21 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp:1005]: this method does not create new tensor. it just updates pointers in a vector.

Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp:1005]: try to re-use already allocated memory for inter. results

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 16 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp:1220]: I will check them

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 1 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp]: We decided to re-use some tensors allocated for intermediate results for the next steps of the computation where it is possible.

File: ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp (hunk 14 above)
Comment (rkazants) [reply ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp]: Fixed.

File: ngraph/test/op_eval/einsum.cpp (hunk 15 above)
Comment (rkazants) [reply ngraph/test/op_eval/einsum.cpp:5]: Added.

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 16 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp:1219]: Fixed.

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 19 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp:653]: I found only one place where it is possible to re-use because shapes of tensors for intermediate results are no changes between iterations.

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 21 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp:1004]: Done.

File: ngraph/core/reference/src/runtime/reference/einsum.cpp (hunk 22 above)
Comment (rkazants) [reply ngraph/core/reference/src/runtime/reference/einsum.cpp]: Done.

**Review by lazarevevgeny — APPROVED** [review]
No major comments.

**Review by ilyachur — APPROVED** [review]
